src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java

```diff
@@ -226,8 +227,9 @@ public TopicPartitionChannel(
       KafkaRecordErrorReporter kafkaRecordErrorReporter,
       SinkTaskContext sinkTaskContext,
       SnowflakeConnectionService conn) {
-    this.streamingIngestClient = Preconditions.checkNotNull(streamingIngestClient);
-    Preconditions.checkState(!streamingIngestClient.isClosed());
+    // TODO: there is no guarantee that the ingestsdkprovider has already been initialized. could
```

Comment:
The following comment states that tasks may be started before/during connector start. Seems like we haven't ran into any issues with that yet, but it is something to think about. Moving the client creation from Task.start to Connector.start may exacerbate this issue.

We already have a boolean in the Connector to indicate when everything has been initialized, perhaps we want to expand the scope of this boolean to here as well. I don't like it, but I'm not sure there are many other good options

https://github.com/snowflakedb/snowflake-kafka-connector/blob/master/src/main/java/com/snowflake/kafka/connector/SnowflakeSinkConnector.java#L65

Reply:
I dont think task can be invoked before a connector start method. 
We can always check for nullability of client in task thread.

Reply:
We do currently check for client nullability, but we throw on it instead of retry. Should I remove this comment about it then?

```
  // Kafka Connect starts sink tasks without waiting for setup in
  // SnowflakeSinkConnector to finish.  
  // This causes race conditions for: config validation, tables and stages
  // creation, etc.
  // Using setupComplete to synchronize
```

Reply:
connector.taskConfigs probably needs to be run prior to task to start, since tasks need a config to start. Due to the setup boolean in taskConfigs, we probably internally can guarantee that. We continue to check for nullability in the task thread, so it should be ok